File: 3.longest-substring-without-repeating-characters/solution.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
		int position[256] = {0};
		int p = 0, q = 0;
		int len = 0;
		while (q < s.length())
		{
			char c = s[q];
			if (position[c] != 0)
			{
				len = max(len, q - p);
				p = position[c];
				q = p;
				memset(position, 0, 256 * sizeof(int));
			} else
			{
				position[c] = ++q;
			}
		}
		return max(len, q - p);
    }
};

File: 3.longest-substring-without-repeating-characters/solution.ts
function lengthOfLongestSubstring(s: string): number {
    const count = new Map<number, number>();
    let start = 0, maxLen = 0;
    for (let i = 0; i < s.length; i++) {
        const char = s.charCodeAt(i);
        const c = count.get(char) ?? 0;
        count.set(char, c+1);
        if (c) {
            let ch: number;
            do {
                ch = s.charCodeAt(start);
                count.set(ch, count.get(ch)! - 1);
                start++;
            } while (ch != s.charCodeAt(i));
        } else {
            maxLen = Math.max(maxLen, i - start + 1);
        }
    }
    return maxLen;
};